contrib/node/src/python/pants/contrib/node/subsystems/eslint_distribution.py

```diff
@@ -0,0 +1,133 @@
+
+class ESLintDistribution(object):
+  """Represents a self-bootstrapping ESLint distribution."""
+
+  ESLINT_DISTRIBUTION_MODULE = 'synthetic-install-eslint-distribution-module'
+
+  class Factory(Subsystem):
+    options_scope = 'eslint-distribution'
+
+    @classmethod
+    def register_options(cls, register):
+      super(ESLintDistribution.Factory, cls).register_options(register)
+      register('--setupdir', advanced=True, fingerprint=True,
+               help='Find the package.json under this dir for installing eslint and plugins.')
+      register('--supportdir', advanced=True, default='bin/eslint',
+               help='Find the ESLint distribution under this dir.')
+      register('--eslint-config', advanced=True, fingerprint=True,
+               help='The path to the global eslint configuration file specifying all the rules')
+      register('--eslint-ignore', advanced=True, fingerprint=True,
+               help='The path to the global eslint ignore path')
+    def create(cls):
+      # NB: create is an instance method used to globally scope this subsystem.
+      options = cls.get_options()
+      global_options = cls.global_instance().get_options()
+      pants_bootstrapdir = global_options.pants_bootstrapdir
+      return ESLintDistribution(options.setupdir, options.supportdir, options.eslint_config,
+                                options.eslint_ignore, pants_bootstrapdir)
+
+  @property
+  def setupdir(self):
+    return self._setupdir
+
+  @property
+  def supportdir(self):
+    return self._supportdir
+
+  @property
+  def eslint_config(self):
+    return self._eslint_config
+
+  @property
+  def eslint_ignore(self):
+    return self._eslint_ignore
+
+  @property
+  def pants_bootstrapdir(self):
+    return self._pants_bootstrapdir
+
+  def __init__(self, setupdir, supportdir, eslint_config, eslint_ignore, pants_bootstrapdir):
+    self._setupdir = setupdir
+    self._supportdir = supportdir
+    self._eslint_config = eslint_config or None
+    self._eslint_ignore = eslint_ignore or None
+    self._pants_bootstrapdir = pants_bootstrapdir
+
+  def _is_workingdir_valid(self, workingdir):
+    dir_exists = os.path.isdir(workingdir)
+    if not dir_exists:
+      return False
+    else:
+      lock_file = os.path.join(workingdir, 'yarn.lock')
+      package_json = os.path.join(workingdir, 'package.json')
+      if not os.path.isfile(package_json):
+        logger.warning('The workingdir: `{}` is missing a package.json'.format(workingdir))
+        return False
+      elif not os.path.isfile(lock_file):
+        logger.warning('A lock file was not found, consider adding a yarn.lock file for '
+                       'deterministic installation.')
+    return True
+
+  def _compare_file_checksums(self, file_a, file_b):
+    if os.path.isfile(file_a) and os.path.isfile(file_b):
+      return self._checksum_md5(file_a) == self._checksum_md5(file_b)
+    return False
+
+  def _checksum_md5(self, filename):
```

Comment:
from pants.base.hash_utils import hash_file should do the trick. It uses SHA1 instead of MD5, but it should be suffice for our purposes. You can also pass in a hashlib.md5() as digest if you really want to use md5.

Reply:
Done